store/searchtest/user_layer.go

```diff
@@ -6,674 +6,891 @@ package searchtest
+		SkipMessage: "Failing when the ListOfAllowedChannels property is empty",
+	},
+	{
+		Name: "Should autocomplete for user using username",
+		Fn:   testAutocompleteUserByUsername,
+		Tags: []string{ENGINE_ALL},
+	},
+	{
+		Name: "Should autocomplete user searching by first name",
+		Fn:   testAutocompleteUserByFirstName,
+		Tags: []string{ENGINE_ALL},
+	},
+	{
+		Name: "Should autocomplete user searching by last name",
+		Fn:   testAutocompleteUserByLastName,
+		Tags: []string{ENGINE_ALL},
+	},
+	{
+		Name: "Should autocomplete for user using nickname",
+		Fn:   testAutocompleteUserByNickName,
+		Tags: []string{ENGINE_ALL},
+	},
+	{
+		Name:        "Should autocomplete for user using email",
+		Fn:          testAutocompleteUserByEmail,
+		Tags:        []string{ENGINE_ALL},
+		Skip:        true,
+		SkipMessage: "Failing for multiple different reasons in the engines",
+	},
+	{
+		Name: "Should be able not to match specific queries with mail",
+		Fn:   testShouldNotMatchSpecificQueriesEmail,
+		Tags: []string{ENGINE_ALL},
+	},
+	{
+		Name: "Should be able to autocomplete a user by part of its username splitted by Dot",
+		Fn:   testAutocompleteUserByUsernameWithDot,
+		Tags: []string{ENGINE_ELASTICSEARCH},
+	},
+	{
+		Name: "Should be able to autocomplete a user by part of its username splitted by underscore",
+		Fn:   testAutocompleteUserByUsernameWithUnderscore,
+		Tags: []string{ENGINE_ELASTICSEARCH},
+	},
+	{
+		Name: "Should be able to autocomplete a user by part of its username splitted by hyphen",
+		Fn:   testAutocompleteUserByUsernameWithHyphen,
+		Tags: []string{ENGINE_ELASTICSEARCH},
+	},
+	{
+		Name: "Should escape the percentage character",
+		Fn:   testShouldEscapePercentageCharacter,
+		Tags: []string{ENGINE_ALL},
+	},
+	{
+		Name: "Should escape the dash character",
+		Fn:   testShouldEscapeUnderscoreCharacter,
+		Tags: []string{ENGINE_ALL},
+	},
+	{
+		Name:        "Should be able to search inactive users",
+		Fn:          testShouldBeAbleToSearchInactiveUsers,
+		Tags:        []string{ENGINE_NONE},
+		Skip:        true,
+		SkipMessage: "Not working",
+	},
+	{
+		Name: "Should be able to search filtering by role",
+		Fn:   testShouldBeAbleToSearchFilteringByRole,
+		Tags: []string{ENGINE_ELASTICSEARCH},
+	},
+	{
+		Name: "Should ignore leading @ when searching users",
+		Fn:   testShouldIgnoreLeadingAtSymbols,
+		Tags: []string{ENGINE_MYSQL, ENGINE_POSTGRES},
+	},
+	{
+		Name: "Should search users in a case insensitive manner",
+		Fn:   testSearchUsersShouldBeCaseInsensitive,
+		Tags: []string{ENGINE_ALL},
+	},
+	{
+		Name: "Should support one or two character usernames and first/last names in search",
+		Fn:   testSearchOneTwoCharUsersnameAndFirstLastNames,
+		Tags: []string{ENGINE_ALL},
+	},
+	{
+		Name: "Should support Korean characters",
+		Fn:   testShouldSupportKoreanCharacters,
+		Tags: []string{ENGINE_ALL},
+	},
+	{
+		Name:        "Should support searching for users containing the term not only starting with it",
```

Comment:
I'd actually like to challenge this requirement and remove it completely. Our approach to autocompletion for users and channels works on a "prefix" basis, so it's matching parts of the fields that start like the term, not that contain them.

The latter approach sounds more like searching rather than autocompleting, so I think we should remove this and add it back if we decide to rethink the approach to the feature, rather than leaving it here as something that we will fix eventually

@wiersgallak what do you think?

Reply:
It's ok for me to remove this test case completely

Reply:
@mgdelacroix @ethervoid How will searching for a user by last name or partial last name be completed then when the username may be like ``kwiersgalla`` or ``katiewiersgalla`` or even ``katie-wiersgalla`` and it is more effective to search with ``wiersgalla`` than `katie` or `k` when first name may return a much larger result set? I believe these were the types of use cases we were trying to solve with this assertion.

Reply:
@wiersgallak I agree on those cases being useful to cover, my point is that right now that's not how the feature works, but something to implement. Leaving the assertion there might be confusing as someone might expect that the feature should work like that and open a bug, when it would be a feature request. I think it's better to add the assertion back if we decide to implement the behavior change.

For the specific cases you noted, if the user has set the last name, the search would match it instead of the username. If the username and the first-last name are not the same, we are splitting the username by `.`, `-` and `_`, so if the username is `katie-wiersgalla`, `wie` would produce a match too. For cases like `kwiersgalla` with no first/last name, `wie` will not produce a match right now

Reply:
@mgdelacroix Makes sense to me.  If we are not currently doing it, we should remove the test case and add it once the feature is implemented.